### PR TITLE
to create a sublogger while getting child of Logger

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
rely on https://github.com/ros2/rcl/pull/921

### CODE REVIEW

File: rclpy/rclpy/impl/rcutils_logger.py [hunk 1]
```diff
@@ -215,15 +215,31 @@ class RcutilsLogger:
...
-            name = self.name + '.' + name
-        return RcutilsLogger(name=name)
+            add_sublogger_flag = _rclpy.rclpy_logging_rosout_add_sublogger(self.name, name)
+            fullname = self.name + '.' + name
+        else:
+            fullname = name
+
+        logger = RcutilsLogger(name=fullname)
+        if self.name and add_sublogger_flag:
+            logger.logger_sublogger_namepair = (self.name, name)
```
Comment (clalancette) [inline rclpy/rclpy/impl/rcutils_logger.py:241]: Like with rclcpp, isn't it enough to just store `self.suffix = suffix` here?

**Review by clalancette — CHANGES_REQUESTED** [review]
One issue, then this generally looks good to me.

File: rclpy/src/rclpy/_rclpy_logging.cpp [hunk 2]
```diff
@@ -172,6 +174,33 @@ rclpy_logging_get_logging_directory()
...
+/// Add a subordinate logger based on a logger
+/// \return True if add sublogger successfully.
+bool
+rclpy_logging_rosout_add_sublogger(const char * logger_name, const char * sublogger_name)
+{
+  rclpy::LoggingGuard scoped_logging_guard;
+  rcl_ret_t rcl_ret = rcl_logging_rosout_add_sublogger(logger_name, sublogger_name);
+  if (RCL_RET_OK != rcl_ret && RCL_RET_SUBLOGGER_ALREADY_EXIST != rcl_ret) {
+    rcutils_reset_error();
+    throw std::runtime_error("Failed to call rcl_logging_rosout_add_sublogger");
```
Comment (fujitatomoya) [inline rclpy/src/rclpy/_rclpy_logging.cpp:186]: Can we add this test case?

File: rclpy/rclpy/impl/rcutils_logger.py [hunk 3]
```diff
@@ -215,15 +215,31 @@ class RcutilsLogger:
...
     def get_child(self, name):
         if not name:
             raise ValueError('Child logger name must not be empty.')
+
+        add_sublogger_flag = False
         if self.name:
             # Prepend the name of this logger
-            name = self.name + '.' + name
-        return RcutilsLogger(name=name)
+            add_sublogger_flag = _rclpy.rclpy_logging_rosout_add_sublogger(self.name, name)
```
Comment (fujitatomoya) [inline rclpy/rclpy/impl/rcutils_logger.py:234]: this comes false means that the subordinate logger already exists in hash map, that actually means that somebody else in the same process space already owns it, so just returning that child logger object without relationship saved in `logger_sublogger_namepair`? right?

**Review by fujitatomoya — APPROVED** [review]
overall lgtm, i got a couple of comments to confirm.

File: rclpy/src/rclpy/_rclpy_logging.cpp (hunk 2 above)
Comment (iuhilnehc-ynos) [reply rclpy/src/rclpy/_rclpy_logging.cpp:186]: Yes, updated.

File: rclpy/rclpy/impl/rcutils_logger.py (hunk 3 above)
Comment (iuhilnehc-ynos) [reply rclpy/rclpy/impl/rcutils_logger.py:234]: Yes.

**Review by fujitatomoya — APPROVED** [review]
lgtm

File: rclpy/rclpy/impl/rcutils_logger.py (hunk 1 above)
Comment (clalancette) [reply rclpy/rclpy/impl/rcutils_logger.py:241]: Never mind this one, I understand now that we need to store it like this.

**Review by clalancette — APPROVED** [review]

**Comment by adityapande-1995** [comment]
Since https://github.com/ros2/rcl/pull/921 is merged, this is just blocked by a release of ``rcl``

**Comment by clalancette** [comment]
> Since [ros2/rcl#921](https://github.com/ros2/rcl/pull/921) is merged, this is just blocked by a release of `rcl`

Which is now done and merged in https://github.com/ros/rosdistro/pull/36283 .  But this is going to take several hours to rebuild (probably overnight).

**Comment by iuhilnehc-ynos** [comment]
Actually, I had already run the full CI for this PR(also including https://github.com/ros2/rclcpp/pull/1717) in https://github.com/ros2/rcl/pull/921#issuecomment-1436319267.
To make the Rpr run sucessfully, I'll wait the `ros-rolling-rcl package` released in http://repo.ros2.org/ubuntu/testing.
After that, I'll trigger the Rpr and re-run a new CI for it.

**Comment by iuhilnehc-ynos** [comment]
@ros-pull-request-builder retest this please

**Comment by iuhilnehc-ynos** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18093/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12625/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18737/)

**Pull request merged**